### PR TITLE
enable loss parallel in SP

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112

In this PR, loss parallel is enabled by default when sequence parallel is enabled.

Below are some empirical results when training for 10 steps, with `sequence_parallel_degree = 4`.
debug_model, with loss parallel disabled
```
Average iter time: 0.1862 seconds
Peak Memory: Reserved 12.9%, Alloc 9.97%, Active: 10.31%
```
debug_model, with loss parallel enabled
```
Average iter time: 0.1055 seconds
Peak Memory: Reserved 3.64%, Alloc 3.18%, Active: 3.25%
```

llama 7B, with loss parallel disabled
```
Average iter time: 6.4379 seconds
Peak Memory: Reserved 34.24%, Alloc 23.64%, Active: 25.89%
```
llama 7B, with loss parallel enabled
```
Average iter time: 6.4778 seconds
Peak Memory: Reserved 27.11%, Alloc 20.51%, Active: 20.51%
```

### CODE REVIEW

File: torchtrain/config_manager.py [hunk 1]
```diff
@@ -185,7 +185,13 @@ def init_args_from_command_line(
...
             type=int,
             default=1,
-            help="Sequence Parallelism degree.  1 means disabled.",
+            help="Sequence Parallelism degree. 1 means disabled.",
+        )
+        parser.add_argument(
+            "--training.enable_loss_parallel",
+            default=True,
+            action="store_true",
+            help="whether to enable loss parallel when sequence parallel is enabled",
```
Comment (awgu) [inline torchtrain/config_manager.py:194]: Is there any downside to enabling loss parallel?

Comment (tianyu-l) [reply torchtrain/config_manager.py:194]: Here's my consideration:
1. Our implementation of loss parallel decomposes large aten ops (log_softmax, nll_loss) to a series torch ops, which creates some computation overhead.
2. loss parallel in general have multiple collectives (three "small" all-reduces) instead of one "big" all-gather. Small vs. big depends on the vocab size, although normal vocab size should always be large enough to justify the memory & communication gain.

For these reasons, I feel we can keep the option but turned on by default. I'm OK with removing this option as well.

File: torchtrain/parallelisms/parallelize_llama.py [hunk 2]
```diff
@@ -144,7 +143,10 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
                 ),
                 "output": ColwiseParallel(
                     input_layouts=Shard(0),
-                    output_layouts=Replicate(),
+                    output_layouts=Shard(-1)
+                    if job_config.training.enable_loss_parallel
+                    else Replicate(),
+                    use_local_output=not job_config.training.enable_loss_parallel,
```
Comment (wanchaol) [inline torchtrain/parallelisms/parallelize_llama.py]: some nits we can follow up latrer: maybe we should pass `job_config.training` to parallelize_llama calls in train.py, so simplify the dict indexing

File: train.py [hunk 3]
```diff
@@ -214,13 +216,15 @@ def main(job_config: JobConfig):
             start_timer.record()
 
             pred = model(input_ids)
-            tok_loss = F.cross_entropy(
-                pred.flatten(0, 1), labels.flatten(0, 1), reduction="none"
+
+            loss_parallel_enabled = (
+                parallel_dims.sp_enabled and job_config.training.enable_loss_parallel
```
Comment (wanchaol) [inline train.py]: can we add a method like `ParallelDims.loss_parallel_enabled`

File: torchtrain/parallelisms/parallelize_llama.py (hunk 2 above)
Comment (tianyu-l) [reply torchtrain/parallelisms/parallelize_llama.py]: ended up using `parallel_dims.loss_parallel_enabled`, so will postpone this for now.
